packages/ifpack2/src/Ifpack2_Details_CrsArrays.hpp

```diff
@@ -0,0 +1,187 @@
+    vals = ScalarArray("", A->getNodeNumEntries());
+    //is there a better lower bound available for this, without allocating a new array for each row?
+    //RowMatrix doesn't have "getMaxNnzPerRow", but that's really what this should be
+    std::vector<Scalar> valsVec(A->getNodeMaxNumRowEntries());
+    Teuchos::ArrayView<Scalar> valsView(valsVec);
+    std::vector<LocalOrdinal> indicesVec(A->getNodeMaxNumRowEntries());
+    Teuchos::ArrayView<LocalOrdinal> indicesView(indicesVec);
+    size_t entry = 0;
+    for(LocalOrdinal row = 0; row < A->getNodeNumRows(); row++)
+    {
+      //Make an ArrayView that views this row's values within vals
+      auto rowEntries = A->getNumEntriesInLocalRow(row);
+      A->getLocalRowCopy(row, indicesView, valsView, rowEntries);
+      //sort the indices, permuting values at the same time
+      Tpetra::sort2(indicesView.getRawPtr(), indicesView.getRawPtr() + rowEntries, valsView.getRawPtr());
+      for(LocalOrdinal i = 0; i < rowEntries; i++)
+      {
+        vals(entry++) = valsView[i];
+      }
+    }
+  }
+
+  static void getRowPtrs(const TRowMatrix* A, OrdinalArray& rowptrs)
+  {
+    auto Acrs = dynamic_cast<const TCrsMatrix*>(A);
+    if(Acrs)
+    {
+      getRowPtrsCrs(Acrs, rowptrs);
+      return;
+    }
+    //Need to allocate new array, then copy in one row at a time
+    //note: actual rowptrs in the CrsMatrix implementation is size_t, but
+    //FastILU needs them as LocalOrdinal so this function provides an OrdinalArray
+    auto nrows = A->getNodeNumRows();
+    rowptrs = OrdinalArray("", nrows + 1);
+    size_t totalEntries = 0;
+    for(LocalOrdinal i = 0; i < A->getNodeNumRows(); i++)
+    {
+      rowptrs[i] = totalEntries;
+      totalEntries += A->getNumEntriesInLocalRow(i);
+    }
+    rowptrs(nrows) = totalEntries;
+  }
+
+  static void getColInds(const TRowMatrix* A, OrdinalArray& colinds)
+  {
+    auto Acrs = dynamic_cast<const TCrsMatrix*>(A);
+    if(Acrs)
+    {
+      getColIndsCrs(Acrs, colinds);
+      return;
+    }
+    //Need to allocate new array, then copy in one row at a time
+    colinds = OrdinalArray("", A->getNodeNumEntries());
+    //is there a better lower bound available for this, without allocating a new array for each row?
+    //RowMatrix doesn't have "getMaxNnzPerRow", but that's really what this should be
+    std::vector<Scalar> valsVec(A->getNodeMaxNumRowEntries());
+    Teuchos::ArrayView<Scalar> valsView(valsVec);
+    std::vector<LocalOrdinal> indicesVec(A->getNodeMaxNumRowEntries());
+    Teuchos::ArrayView<LocalOrdinal> indicesView(indicesVec);
+    size_t entry = 0;
+    for(LocalOrdinal row = 0; row < A->getNodeNumRows(); row++)
+    {
+      //Make an ArrayView that views this row's values within vals
+      auto rowEntries = A->getNumEntriesInLocalRow(row);
+      A->getLocalRowCopy(row, indicesView, valsView, rowEntries);
+      //sort indices
+      std::sort(indicesView.getRawPtr(), indicesView.getRawPtr() + rowEntries);
```

Comment:
@jhux2 I do a dynamic cast check for CrsMatrix and in that case just pull the 3 CRS arrays directly from the Kokkos::CrsMatrix and Kokkos::StaticCrsGraph. I think that's always valid because at this point the matrix should be fillComplete.